### PR TITLE
Upgrade core components to .NET 6

### PR DESCRIPTION
**Pull request by bastbu**
# PR for issue #817

## What is being addressed

With this PR we upgrade our core components (LNS, DecoderSample, Provisioning Tool) to .NET 6. This includes the usage of the Azure Functions v4 runtime, which is generally available.

## How is this addressed

- We update the SDK version in the `global.json`
- We update the target versions of the projects
- We update the CI/Docker image SDKs
- We handle code analysis and deprecation warnings
    - Only some CA warnings to use structured logging were addressed, while we push out solving other instances of that warning to #829 

After an initial review: this PR will only be merged during off-hours, where we can upgrade our E2E Function to runtime v4 and ensure that the E2E tests are running.

Follow-up tasks:
- Ensure documentation is up to date
- Update Code Analysis to Analysis Level 6.

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LoRaMessage/LoRaPayload.cs [hunk 1]
```diff
@@ -115,19 +115,17 @@ public static byte[] CalculateKey(LoRaPayloadKeyType keyType, byte[] appnonce, b
...
 #pragma warning disable CA5358 // Review cipher mode usage with cryptography experts
-                // Cipher is part of the LoRaWAN specification
-                Mode = CipherMode.ECB,
+            // Cipher is part of the LoRaWAN specification
+            aes.Mode = CipherMode.ECB;
 #pragma warning restore CA5358 // Review cipher mode usage with cryptography experts
-                Padding = PaddingMode.None
-            };
+            aes.Padding = PaddingMode.None;
+            aes.IV = new byte[16];
```
Comment (Mandur) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LoRaMessage/LoRaPayload.cs:125]: what's this ? IV?

Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/LoRaMessage/LoRaPayload.cs:125]: It was on L130 previously

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs [hunk 2]
```diff
@@ -254,29 +253,17 @@ public int CompareTo(ApiVersion other)
             Name == version.Name &&
             IsKnown == version.IsKnown;
 
-        public static bool operator <(ApiVersion value1, ApiVersion value2)
-        {
-            if (value1 is null) throw new ArgumentNullException(nameof(value1));
-            return value1.CompareTo(value2) < 0;
-        }
+        public static bool operator <(ApiVersion value1, ApiVersion value2) =>
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs:256]: I think the whole comparison stuff might not be fully tested or even flawed so why not just remove it instead of adding more fixes and tests. It doesn't seem to be used anywhere except in `SupportsVersion`, where it can be all in-lined. This could be done in a separate PR that's merged into this.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs [hunk 3]
```diff
@@ -177,7 +177,7 @@ async Task LoopAsync()
 
                     var prunedKeys = Prune();
 
-                    this.logger?.LogDebug($"Pruned closed WebSocket connections: {string.Join(",", prunedKeys)}");
+                    this.logger?.LogDebug("Pruned closed WebSocket connections: {PrunedKeys}", string.Join(",", prunedKeys));
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs:180]: What's the reason for this change? Do we need structured logging here?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs [hunk 4]
```diff
@@ -22,6 +22,9 @@ namespace LoRaWan.NetworkServer.BasicsStation.Processors
 
     internal class LnsProtocolMessageProcessor : ILnsProtocolMessageProcessor
     {
+        private static readonly Action<ILogger, string, string, Exception> logReceivedMessage =
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:25]: Static read-only fields generally use Pascal casing because they are considered like constants, albeit run-time ones.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs [hunk 5]
```diff
@@ -72,10 +75,10 @@ internal class LnsProtocolMessageProcessor : ILnsProtocolMessageProcessor
 #pragma warning restore CA1508 // Avoid dead conditional code
                 {
                     // This can happen if the basic station client is losing connectivity
-                    this.logger.LogDebug(ex, ex.Message);
+                    this.logger.LogDebug(ex, "{Message}", ex.Message);
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:78]: This looks very strange where the whole message is a placeholder?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs [hunk 6]
```diff
@@ -15,6 +15,10 @@ namespace LoRaWan.NetworkServer.BasicsStation
 
     internal class DownstreamSender : IPacketForwarder
     {
+        private static readonly Action<ILogger, StationEui, string, Exception> logSendingMessage =
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs:18]: Static read-only fields generally use Pascal casing because they are considered like constants, albeit run-time ones.

File: Directory.Build.props [hunk 7]
```diff
@@ -4,47 +4,48 @@
       <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
       <AnalysisMode>AllEnabledByDefault</AnalysisMode>
       <AnalysisLevel>5.0</AnalysisLevel>
-      <LangVersion>9.0</LangVersion>
+      <LangVersion>10.0</LangVersion>
```
Comment (atifaziz) [inline Directory.Build.props:7]: 🎉

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs (hunk 5 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:78]: What alternative would you suggest - suppress the warning?

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs (hunk 2 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs:256]: I added tests for the comparison before addressing the warnings - did I miss a case there? Curious to learn, why should it be flawed, is it because of comparing to `null`?

File: Samples/DecoderSample/SensorDecoderModule.csproj [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Samples/DecoderSample/SensorDecoderModule.csproj:4]: Generally speaking we don't need to follow the same SDK upgrades for the samples/CLI as with the Core modules. For example, the cayenne module hasn't been updated since .NET Core 2.1, hence I thought that I will share the target framework only where necessary.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs (hunk 3 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs:180]: Yes. A new code warning disallows log message templates that change with each invocation, i.e. forces us to use structured logging.

File: Tests/Unit/LoraKeysManagerFacade/FunctionBundlerTest.cs [hunk 9]
```diff
@@ -248,7 +248,7 @@ public async Task FunctionBundler_Multi()
                 }));
             }
 
-            Task.WaitAll(tasks.ToArray());
+            await Task.WhenAll(tasks);
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Tests/Unit/LoraKeysManagerFacade/FunctionBundlerTest.cs]: Addressed in 72aad1d0197104911e0677cb7b9c39c70287320f

File: Tests/Unit/LoRaWan.Tests.Unit.csproj [hunk 10]
```diff
@@ -1,6 +1,6 @@
-﻿<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk">
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Tests/Unit/LoRaWan.Tests.Unit.csproj]: 531534b8407351139a02fced91be9056b993af19

File: Tools/Cli-LoRa-Device-Provisioning/Cli-LoRa-Device-Provisioning.csproj [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFramework>net6.0</TargetFramework>
```
[the remark this answers is not in the document]
Comment (bastbu) [reply Tools/Cli-LoRa-Device-Provisioning/Cli-LoRa-Device-Provisioning.csproj]: Addressed in d77c790ccd3d7793c7a3f8108531f5d205269b97

File: Samples/DecoderSample/SensorDecoderModule.csproj (hunk 8 above)
[the remark this answers is not in the document]
Comment (bastbu) [reply Samples/DecoderSample/SensorDecoderModule.csproj]: Addressed in d77c790ccd3d7793c7a3f8108531f5d205269b97

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs (hunk 3 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/WebSocketWriterRegistry.cs]: Reverted this with a29c611156a5b208419fac20e931f8f8661cce72. I suppressed the warning for now, since the merge of #842 generated many more of these instances. We will address this as part of a separate task (linked in the description). Since I already fixed some of these warnings in an older version of the branch, I will not revert those.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs (hunk 5 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs]: Reverted this with a29c611156a5b208419fac20e931f8f8661cce72. I suppressed the warning for now, since the merge of #842 generated many more of these instances. We will address this as part of a separate task (linked in the description). Since I already fixed some of these warnings in an older version of the branch, I will not revert those.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs (hunk 2 above)
Comment (atifaziz) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs:256]: > Curious to learn, why should it be flawed…?

Because it seems wrong to me that the equality implementation considers the 3 properties of a version (`Version`, `Name` and `IsKnown`) yet comparison only cares about `Version`, and what's more, `CompareTo` initially depends on equality/`Equals`. This tells me there's some indecision about what fundamentally constitutes as the identity or value of the type and this can lead to surprises or subtle bugs.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs [hunk 12]
```diff
@@ -254,29 +253,17 @@ public int CompareTo(ApiVersion other)
...
+            value1 is { } v1 ? value1.CompareTo(value2) < 0 : value2 is not null;
```
Comment (atifaziz) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs:257]: Did you mean to use `v1` here?

```suggestion
            value1 is { } v1 ? v1.CompareTo(value2) < 0 : value2 is not null;
```

**Review by atifaziz — APPROVED** [review]
Apart from the comments on the value of the `ApiVersion` comparison operators and its comparison implementation being flawed, this LGTM.

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs (hunk 12 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs:257]: Good catch, I will inline this whole logic on the call-site though to get rid of the wrong implementation

Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs]: Addressed this in ebe9bde78806fbf0c0384f25763d9cbbfcdd14a4

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs (hunk 2 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/CommonAPI/ApiVersion.cs]: Addressed in ebe9bde78806fbf0c0384f25763d9cbbfcdd14a4
